Optimize frequent combinations of relational operators

bjorng opened the pull request.

bjorng committed:
compiler: Turn is_lt into is_ge if possible

This is not an optimization in itself but it reduces the number
of permuations of adjacent relational operators and thus will
simplify instruction combination optimizations for the JIT.

bjorng committed:
jit: Optimize relational guard tests

Optimize the generation of native code for common combinations of
relational operators in guards.

jhogberg commented on erts/emulator/beam/jit/arm/beam_asm.hpp:
```diff
@@ -739,6 +739,15 @@ class BeamAssembler : public ErrorHandler {
     }
 
+    void cmp(arm::Gp src, int64_t val) {
+        if (0 <= val && val <= 0xFFF) {
+            a.cmp(src, imm(val));
+        } else {
```
```suggestion
        if (Support::isUInt12(val)) {
            a.cmp(src, imm(val));
        } else if (Support::isUInt12(-val)) {
            a.cmn(src, imm(-val));
        } else {
```

bjorng committed:
Slightly optimize the is_eq_exact and is_ne_exact instructions

The `cmp_arg()` helper would generate an unnecessary move to the
SUPER_TMP register if the RHS argument was an X register backed by a
CPU register. For example:

    mov x14, x27
    cmp x26, x14

This code can be simplified to:

    cmp x26, x27

bjorng committed:
Update primary bootstrap
